Source/WebCore/platform/audio/cocoa/PlatformRawAudioDataCocoa.cpp

```diff
@@ -0,0 +1,320 @@
+{
+    return m_sample->sizeInBytes();
+}
+
+bool PlatformRawAudioDataCocoa::isInterleaved() const
+{
+    return m_description.isInterleaved();
+}
+
+static std::variant<Vector<std::span<uint8_t>>, Vector<std::span<int16_t>>, Vector<std::span<int32_t>>, Vector<std::span<float>>> planesOfSamples(AudioSampleFormat format, const WebAudioBufferList& list, size_t samplesOffset)
+{
+    WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    switch (format) {
+    case AudioSampleFormat::U8:
+    case AudioSampleFormat::U8Planar:
+        return Vector<std::span<uint8_t>> { list.bufferCount(), [&](auto index) {
+            return unsafeMakeSpan(static_cast<uint8_t*>(list.buffer(index)->mData) + samplesOffset, list.buffer(index)->mDataByteSize - samplesOffset);
+        } };
+    case AudioSampleFormat::S16:
+    case AudioSampleFormat::S16Planar:
+        return Vector<std::span<int16_t>> { list.bufferCount(), [&](auto index) {
+            return unsafeMakeSpan(static_cast<int16_t*>(list.buffer(index)->mData) + samplesOffset, list.buffer(index)->mDataByteSize / sizeof(int16_t) - samplesOffset);
+        } };
+    case AudioSampleFormat::S32:
+    case AudioSampleFormat::S32Planar:
+        return Vector<std::span<int32_t>> { list.bufferCount(), [&](auto index) {
+            return unsafeMakeSpan(static_cast<int32_t*>(list.buffer(index)->mData) + samplesOffset, list.buffer(index)->mDataByteSize / sizeof(int32_t) - samplesOffset);
+        } };
+    case AudioSampleFormat::F32:
+    case AudioSampleFormat::F32Planar:
+        return Vector<std::span<float>> { list.bufferCount(), [&](auto index) {
+            return unsafeMakeSpan(static_cast<float*>(list.buffer(index)->mData) + samplesOffset, list.buffer(index)->mDataByteSize / sizeof(float) - samplesOffset);
+        } };
+    }
+    WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    RELEASE_ASSERT_NOT_REACHED();
+    return Vector<std::span<uint8_t>> { };
+}
+
+void PlatformRawAudioData::copyTo(std::span<uint8_t> destination, AudioSampleFormat destinationFormat, size_t planeIndex, std::optional<size_t> frameOffset, std::optional<size_t>, unsigned long copyElementCount)
+{
+    auto& audioData = downcast<PlatformRawAudioDataCocoa>(*this);
+
+    auto sourceFormat = format();
+    WebAudioBufferList sourceList(audioData.m_description, audioData.sampleBuffer());
+    bool destinationIsInterleaved = isAudioSampleFormatInterleaved(destinationFormat);
+
+    if (audioSampleElementFormat(sourceFormat) == audioSampleElementFormat(destinationFormat)) {
+        if (numberOfChannels() == 1 || (audioData.isInterleaved() && destinationIsInterleaved)) {
+            // Simplest case.
+            ASSERT(!planeIndex);
+            size_t frameOffsetInBytes = frameOffset.value_or(0) * audioData.m_description.bytesPerFrame();
+            auto source = unsafeMakeSpan(static_cast<uint8_t*>(sourceList.buffer(0)->mData) + frameOffsetInBytes, sourceList.buffer(0)->mDataByteSize - frameOffsetInBytes);
```

Comment:
Are we sure we validate enough all values here at the caller site?
Ditto below.

Also, WebAudioBufferList could provide a `span<const uint8_t> bufferAsSpan(int)` so that we would use subspace here.
Or WebAudioBufferList would provide a copy method itself.

Reply:
The validation is done in Compute Copy Element Count ; I triple checked that code, but dealing with flat raw memory buffer there's always a risk of something in there. The spec itself doesn't make it safe.

But the copyElementCount account for the destination size. I'll try to make the flow a bit less convoluted and clear that it's safe